modules/bim/db/migrate/20201105154216_seed_custom_style_with_bim_theme.rb

```diff
@@ -0,0 +1,111 @@
+#-- encoding: UTF-8
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2020 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2017 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+# This migration cleans up messed up themes. Sometimes in the past
+# the BIM theme was not set where it should have been set.
+class SeedCustomStyleWithBimTheme < ActiveRecord::Migration[6.0]
+
+  def up
+    # When
+    #   migrating BIM instances
+    #     that did not have any custom styles OR
+    #       that do not have any design colors set and no custom logo/touch-icon/favicon
+    #       (this basically means that no theme actually got applied)
+    # then
+    #   add a custom style with the BIM theme set. This will write the theme's colors
+    #   as DesignColor entries to the DB which is necessary for the theme to actually
+    #   have an effect.
+    if OpenProject::Configuration.bim? &&
+       (CustomStyle.current.nil? ||
+           (DesignColor.count == 0 &&
+               CustomStyle.current.favicon.nil? &&
+               CustomStyle.current.logo.nil? &&
+               CustomStyle.current.touch_icon.nil?))
+      seed_bim_theme
+    end
+  end
+
+  def down
+    # nop
+  end
+
+  private
+
+  def seed_bim_theme
+    CustomStyle.transaction do
+      set_logo
+      set_colors
+      set_theme
+
+      custom_style.save!
+    end
```

Comment:
This seems a bit complicated. It boils down to two operations:
* Adding the `DesignColor`s
* Adding or updating the `CustomStyle`

Calls to two methods (`set_colors` and `set_style` or something like this) should be in this method then. The `custom_style.save!` really confused me, as it does not appear in the method before and I then have to figure out that the methods before have side effects on the custom_style (but not the `set_colors` method). 

The `set_style` method could be written like this:

```
custom_style = (CustomStyle.current || CustomStyle.new)
custom_style.attributes = { theme: theme[:theme], theme_logo: theme[:logo] }
custom_style.save!
```